app/src/main/java/com/infomaniak/mail/ui/main/move/MoveViewModel.kt

```diff
@@ -49,38 +49,70 @@ class MoveViewModel @Inject constructor(
     private val messageUid inline get() = savedStateHandle.get<String?>(MoveFragmentArgs::messageUid.name)
     private val threadsUids inline get() = savedStateHandle.get<Array<String>>(MoveFragmentArgs::threadsUids.name)!!
 
-    var filterResults: MutableLiveData<List<Folder>> = MutableLiveData()
+    var filterResults: MutableLiveData<Pair<List<Folder>, String>> = MutableLiveData()
 
-    fun cancelSearch() {
-        filterJob?.cancel()
-    }
-
-    fun getFolderIdAndCustomFolders() = liveData(ioCoroutineContext) {
+    fun getCurrentFolderAndAllFolders() = liveData(ioCoroutineContext) {
 
-        val folderId = messageUid?.let { messageController.getMessage(it)!!.folderId }
+        val currentFolderId = messageUid?.let(messageController::getMessage)?.folderId
             ?: threadController.getThread(threadsUids.first())!!.folderId
 
-        val customFolders = folderController.getCustomFolders().getCustomMenuFolders()
+        val folders = folderController.getMoveFolders().flattenFolderChildren().addDividerToFirstCustomFolder()
 
-        emit(folderId to customFolders)
+        emit(folders to currentFolderId)
     }
 
-    fun filterFolders(query: String, folders: List<Folder>, shouldDebounce: Boolean) = viewModelScope.launch(ioCoroutineContext) {
+    private fun List<Folder>.addDividerToFirstCustomFolder(): List<Folder> {
+        var needsToAddDivider = true
+        return map { folder ->
+            folder.apply {
+                if (needsToAddDivider && isRootAndCustom) {
+                    needsToAddDivider = false
+                    shouldDisplayDivider = true
+                }
+            }
+        }
+    }
+
+    fun filterFolders(
+        query: String,
+        folders: List<Folder>,
+        currentFolderId: String,
+        shouldDebounce: Boolean,
+    ) = viewModelScope.launch(ioCoroutineContext) {
         filterJob?.cancel()
         filterJob = launch {
             if (shouldDebounce) {
                 delay(FILTER_DEBOUNCE_DURATION)
                 ensureActive()
             }
-            val filteredFolders = folders.filter { folder ->
+
+            var isFirstRootAndCustom = true
+            val filteredFolders = folders.mapNotNull { folder ->
                 val folderName = folder.role?.folderNameRes?.let(appContext::getString) ?: folder.name
-                folderName.standardize().contains(query.standardize())
+                val isFound = folderName.standardize().contains(query.standardize())
+                if (isFound) {
+                    folder.clone().apply {
+                        shouldDisplayDivider = if (isRootAndCustom && isFirstRootAndCustom) {
```

Comment:
`isRootAndCustom` is heavier to compute, we might as well do it this way since it doesn't impact anything else also this should be pretty easy for processors optimizations not that it matter much

```suggestion
                        shouldDisplayDivider = if (isFirstRootAndCustom && isRootAndCustom) {
```